pkg/addonmanager/controllers/agentdeploy/hosting_controller.go

```diff
@@ -0,0 +1,346 @@
+				accessor, _ := meta.Accessor(obj)
+				if accessor.GetLabels() == nil {
+					return false
+				}
+
+				addonName, ok := accessor.GetLabels()[constants.AddonLabel]
+				if !ok {
+					return false
+				}
+				addonNamespace, ok := accessor.GetLabels()[constants.AddonNamespaceLabel]
+				if !ok {
+					return false
+				}
+
+				if _, ok := c.agentAddons[addonName]; !ok {
+					return false
+				}
+				if accessor.GetName() != constants.DeployHostingWorkName(addonNamespace, addonName) {
+					return false
+				}
+				return true
+			},
+			workInformers.Informer(),
+		).
+		WithSync(c.sync).ToController("addon-hosting-deploy-controller")
+}
+
+func (c *addonHostingDeployController) sync(ctx context.Context, syncCtx factory.SyncContext, key string) error {
+	klog.V(4).Infof("Reconciling addon hosting deploy %q", key)
+	klog.Infof("Reconciling addon hosting deploy %q", key)
+
+	clusterName, addonName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is not in format: namespace/name
+		return nil
+	}
+
+	agentAddon, ok := c.agentAddons[addonName]
+	if !ok {
+		return nil
+	}
+
+	// Get ManagedCluster
+	managedCluster, err := c.managedClusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if !managedCluster.DeletionTimestamp.IsZero() {
+		// managed cluster is deleting, do nothing
+		return nil
+	}
+
+	managedClusterAddon, err := c.managedClusterAddonLister.ManagedClusterAddOns(clusterName).Get(addonName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
+
+	installMode, hostingClusterName := constants.GetHostedModeInfo(managedClusterAddon.GetAnnotations())
+	if installMode != constants.InstallModeHosted {
+		// the deploy mode changed from hosted to default, cleanup the hosting resources
+		return c.cleanup(ctx, managedClusterAddonCopy, "")
+	}
+
+	// Get Hosting Cluster, check whether the hosting cluster is a managed cluster of the hub
+	// TODO: check whether the hosting cluster of the addon is the same hosting cluster of the klusterlet
+	hostingCluster, err := c.managedClusterLister.Get(hostingClusterName)
+	if errors.IsNotFound(err) {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    constants.HostingClusterValidity,
+			Status:  metav1.ConditionFalse,
+			Reason:  constants.HostingClusterValidityReasonInvalid,
+			Message: fmt.Sprintf("hosting cluster %s is not a managed cluster of the hub", hostingClusterName),
+		})
+		if pErr := utils.PatchAddonCondition(
+			ctx, c.addonClient, managedClusterAddonCopy, managedClusterAddon); pErr != nil {
+			return fmt.Errorf("failed to update managedclusteraddon status: %v; the err should be %v", pErr, err)
+		}
+
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
```

Comment:
we can do it later so do not reconcile so fast.

Reply:
you don't quite understand, you mean we just set condition, but don't patch condition immediately?

Reply:
yes